### PR TITLE
Authenticate connection IDs

### LINKED ISSUE #3429: A fixed transport parameter profile is legitimate
In the description of active_connection_id_limit, we require that the server make the value of the transport parameter conditional on the choice of connection ID from the client.

In the past, we have had the principle that either endpoint not condition its choice of transport parameter values based on anything from its peer.  We should maintain this principle throughout.

This means changing the definition of active_connection_id_limit;  @larseggert had a sense that there was another.

### LINKED ISSUE #3439: Authenticating connection IDs
@ad-l has noted in the past that we don't authenticate the connection IDs that are exchanging during the handshake.  This complicates analysis of header protection (for details, see the upcoming paper).  We already authenticate the connection ID that the client first uses if there is a Retry packet sent, but we don't otherwise authenticate the connection IDs that are signaled in the Source Connection ID field.

The question here is whether we should be more careful about authenticating these.  

@nharper also points out in #3429 that the ODCID transport parameter is an odd duck.  If we use transport parameters to authenticate connection IDs, we might be better served by restoring a fixed header to the transport parameters so that we can more cleanly address that issue.

### PR DESCRIPTION
**Pull request by martinthomson**
This defines a new transport parameter and new rules for the server when it generates the Source Connection ID for Initial packets.

See [discussion on the issue](https://github.com/quicwg/base-drafts/issues/3439#issuecomment-594320754) for rationale and background.

Closes #3439.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1453,13 +1453,21 @@ ID values as the client's first Initial.
...
-establishment, once in response to a Retry and once in response to the first
-Initial packet from the server. Once a client has received an Initial packet
+for subsequent packets, including any 0-RTT packets.  A client changes the
+Destination Connection ID field if it receives a Retry packet.  A client always
+changes the Destination Connection ID field in response to the first Initial
+packet received from a server.  Once a client has received an Initial packet
 from the server, it MUST discard any packet it receives with a different Source
 Connection ID.
 
+Each endpoint includes the connection ID it includes in Initial packets in the
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: I think this text should be more specific about which connection ID it's talking about (the source connection ID). Also, a sentence spelling out how validation is performed can't hurt.

Comment (martinthomson) [reply draft-ietf-quic-transport.md]: Thanks David.  Fixed; and you get two sentences.

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1453,13 +1453,24 @@ ID values as the client's first Initial.
...
-establishment, once in response to a Retry and once in response to the first
-Initial packet from the server. Once a client has received an Initial packet
+for subsequent packets, including any 0-RTT packets.  A client changes the
+Destination Connection ID field if it receives a Retry packet.  A client always
+changes the Destination Connection ID field in response to the first Initial
+packet received from a server.  Once a client has received an Initial packet
 from the server, it MUST discard any packet it receives with a different Source
 Connection ID.
 
+Each endpoint includes the Source Connection ID it includes in Initial packets
```
Comment (chris-wood) [inline draft-ietf-quic-transport.md]:
```suggestion
Each endpoint includes its Initial packet Source Connection ID 
```

Comment (martinthomson) [reply draft-ietf-quic-transport.md]: Thanks.  I've tried to avoid that sort of train of nouns in the past.  I have an alternative for you.  More words again :(, but maybe clearer.

Comment (chris-wood) [reply draft-ietf-quic-transport.md]: No worries! Just a suggestion in reaction to redundant "includes."

**Comment by DavidSchinazi** [comment]
I agree that this is the right fix. However, I think I would prefer to allow the server to reuse the client's initial DCID when sending the first server initial when there are no retries involved. From a theoretical standpoint, requiring the server to change the CID makes the one offered by the client useless which seems silly - but that is admittedly a very weak argument. From a practical standpoint, that would allow us to deploy this change faster.

**Comment by ianswett** [comment]
David is correct that that change would speed our deployment by quite a bit.

I'm more concerned about raising the bar too high for implementers that have very simple load balancing infrastructure, so I think this deserves an email to the list if you're willing, @martinthomson

**Comment by kazuho** [comment]
IMO I think it is a good idea to discourage servers from using client-generated data for load balancing etc., even though we might agree that we do not need to forbid that.

Maybe SHOULD NOT?

**Comment by DavidSchinazi** [comment]
I'm not sure we should be discouraging servers from using client-generated connection IDs. We've been doing that for years and I'm not sure there is a credible attack here.

**Comment by martinthomson** [comment]
Discuss this on the issue please.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1453,13 +1453,24 @@ ID values as the client's first Initial.
...
+packets it sends in the handshake_connection_id transport parameter; see
+{{transport-parameter-definitions}}.  Each endpoint validates that the value
+received from the peer is identical to the value of the transport parameter.
+Absense of the handshake_connection_id transport parameter or a mismatch in
+values MUST be treated as a connection error of type PROTOCOL_VIOLATION.  When
+sending a Retry packet or the first Initial packet, a server MUST select values
+for the Source Connection ID field that differ from the values the client
+includes in the Destination Connection ID field.  These measures ensure that the
+choice of connection ID cannot be influenced by an attacker.
+
```
Comment (huitema) [inline draft-ietf-quic-transport.md]: I have some difficulty parsing that in the retry scenario. is "the Source Connection ID from Initial
packet" the value found in the retry packet sent by the server, or is the source connection ID in the first initial packet of the server? Currently, there is no requirement that these two be the same value.

**Comment by martinthomson** [comment]
Hi everyone, I've redone this based on discussion.  Please take another look.

@huitema, I don't think that this benefits from a description of an attack.  This is not a change made in response to an attack, but in response to a request to make the protocol easier to analyze for the purposes of proving its security.

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1475,6 +1475,44 @@ lifetime of a connection, especially in response to connection migration
...
+* absence of the original_connection_id transport parameter from the server,
+
+* absence of the retry_connection_id transport parameter from the server after
+  receiving a Retry packet,
+
+* presence of the retry_connection_id transport parameter when no Retry packet
+  was received, or
+
+* a mismatch between values of these transport parameters and the Destination
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: I think you mean `a mismatch between values of these transport parameters and the respective Connection ID fields an endpoint used.` since it doesn't just protect destination

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -1522,9 +1560,10 @@ specify whether they MUST, MAY, or MUST NOT be stored for 0-RTT. A client need
 not store a transport parameter it cannot process.
 
 A client MUST NOT use remembered values for the following parameters:
-original_connection_id, preferred_address, stateless_reset_token, and
-ack_delay_exponent. The client MUST use the server's new values in the
-handshake instead, and absent new values from the server, the default value.
+ack_delay_exponent, handshake_connection_id, original_connection_id,
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: Shouldn't `original_connection_id` stay in this list? And what about `handshake_connection_id`?

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -4357,13 +4395,15 @@ A client MUST NOT reset the packet number for any packet number space after
 processing a Retry packet; {{packet-0rtt}} contains more information on this.
 
 A server acknowledges the use of a Retry packet for a connection using the
-original_connection_id transport parameter (see
-{{transport-parameter-definitions}}).  If the server sends a Retry packet, it
+retry_connection_id transport parameter; see
+{{transport-parameter-definitions}}. If the server sends a Retry packet, it
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: We may want to rephrase this to more clearly state that `original_connection_id` is always filled in by the server

File: draft-ietf-quic-transport.md [hunk 7]
```diff
@@ -4572,9 +4612,8 @@ original_connection_id (0x00):
 
 : The value of the Destination Connection ID field from the first Initial packet
   sent by the client.  This transport parameter is only sent by a server.  This
-  is the same value sent in the "Original Destination Connection ID" field of a
-  Retry packet (see {{packet-retry}}).  A server MUST include the
-  original_connection_id transport parameter if it sent a Retry packet.
+  is the same value sent in the "Original Destination Connection ID" field that
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: Thanks for catching that, my bad for missing it.

File: draft-ietf-quic-transport.md [hunk 8]
```diff
@@ -4742,6 +4781,22 @@ active_connection_id_limit (0x0e):
   When a zero-length connection ID is being used, the active_connection_id_limit
   parameter MUST NOT be sent.
 
+handshake_connection_id (0x0f):
+
+: The value that the endpoint included in the Source Connection ID field of the
+  first Initial packet it sends during the handshake. Endpoints MUST validate
+  that this transport parameter is present and that it matches the value that
+  was received in Initial packets. Authenticating this value ensures that an
+  attacker is unable to influence the selection of connection IDs during the
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md]: to avoid jinxing things, how about `Authenticating this value mitigates an attacker's ability to influence the selection of connection IDs during the handshake.`

Comment (martinduke) [reply draft-ietf-quic-transport.md]: What if some of these CIDs are zero length? Do we omit them?

Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md]: My read of the zero-length CID case is that the TP MUST be included, and be zero-length. Perhaps we should add a sentence to clarify that.

Comment (martinthomson) [reply draft-ietf-quic-transport.md]: Done! Thanks both.

As for the "mitigates" formulation, I think that we are safe in making claims, but I realize that we don't need to.  I've removed a lot of text from the transport parameter definitions and replaced it with a pointer to the new section.

File: draft-ietf-quic-transport.md [hunk 9]
```diff
@@ -1475,6 +1475,48 @@ lifetime of a connection, especially in response to connection migration
...
 
+## Authenticating Connection IDs {#cid-auth}
+
+The choice each endpoint makes about connection IDs during the handshake is
+authenticated by including all values in transport parameters; see
+{{transport-parameters}}. This ensures that all connection IDs are authenticated
+by the cryptographic handshake.
+
+Each endpoint includes the value of the Source Connection ID field from Initial
+packets it sends in the handshake_connection_id transport parameter; see
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: One could read that this Connection ID is only used on Initial packets.  If so, I think initial_connection_id would be a better TP name.  But I think this CID can be used used on Initial and 0-RTT from the client and all Initial, Handshake and 1-RTT from the server?

File: draft-ietf-quic-transport.md [hunk 10]
```diff
@@ -1475,6 +1475,48 @@ lifetime of a connection, especially in response to connection migration
...
+  receiving a Retry packet,
+
+* presence of the retry_connection_id transport parameter when no Retry packet
+  was received, or
+
+* a mismatch between values received from a peer in these transport parameters
+  and the value sent in the corresponding Destination Connection ID fields of
+  Initial packets.
+
+If a zero-length connection IDs is selected, the corresponding transport
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]:
```suggestion
If a zero-length connection ID is selected, the corresponding transport
```

**Comment by huitema** [comment]
@ianswett I have implement it in picoquic as such:

1) Always send the handshake CID. Not an interop issue since implementations are supposed to ignore unknown CID

2) Send the retry CID when also sending the ODCID.

3) Verify that the client is not sending Retry CID

4) If the protocol version is newer than "draft 27", check that the handshake CID is present, and that the retry CID is present if required. Then check that the values are what is expected.

5) When verifying the retry token, check that the retry CID sent by the client is what is expected.

That version of picoquic is used in the interop runner, and I have not noticed any issue related to CID checks. But then the tests are bypassed with draft-27 versions. I suppose I could split the TP test into verifying that the CID are present if the version is newer than draft-27, and then check that the CID have the expected value if they are present. That would allow for interop testing with draft 27.

File: draft-ietf-quic-transport.md (hunk 9 above)
Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md]: I think it's called `handshake_connection_id` because it is the connection ID used during the handshake, I think it's a reasonable name

**Comment by DavidSchinazi** [comment]
I like @huitema's deployment plan. I think it would be valuable to get deployment experience of these new transport parameters in draft-27 so that it doesn't break interop. Perhaps we should add a temporary subsection to the spec on how to this, with a note that it'll be removed before publication. It could say something like this:
1. If using draft-28 or later, these transport parameters MUST be sent, and received ones MUST be validated.
2. If using draft-27 or earlier, these transport parameters SHOULD be sent, and upon receipt:
    - if `handshake_connection_id` is present, then validate using draft-28 rules
    - else validate using draft-27 rules

Thoughts?

**Comment by huitema** [comment]
@DavidSchinazi in practice, you can only publish new spec as draft-28. So what you want is not a new draft, but maybe a note on the implementers' slack. By the way, I did change my implementation to check the values if they are sent.

**Comment by DavidSchinazi** [comment]
@huitema That sounds good to me, or perhaps on the implementer's wiki

**Comment by huitema** [comment]
@DavidSchinazi, @ianswett I just added a wiki page on [testing this PR](https://github.com/quicwg/base-drafts/wiki/Testing-the-authentication-of-connection-ID-before-draft-28).

File: draft-ietf-quic-transport.md [hunk 11]
```diff
@@ -1475,6 +1475,48 @@ lifetime of a connection, especially in response to connection migration
...
+Each endpoint includes the value of the Source Connection ID field from Initial
+packets it sends in the handshake_connection_id transport parameter; see
+{{transport-parameter-definitions}}. A server includes the Destination
+Connection ID field it receives in the original Initial packet from the client
+in the original_connection_id transport parameter. After sending a Retry
+packet, a server also includes the Source Connection ID field from the Retry
+packet in the retry_connection_id transport parameter.
+
+Each endpoint ensures that the values of transport parameters matches the
+values. The values provided by a peer for these transport parameters MUST match
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md]: This sentence feels like it ends prematurely.  Perhaps...
```suggestion
Each endpoint ensures that the values of the peer's transport parameters
match the values the endpoint sent in its packet headers. The values
provided by a peer for these transport parameters MUST match the
```

File: draft-ietf-quic-transport.md (hunk 9 above)
Comment (janaiyengar) [reply draft-ietf-quic-transport.md:1486]: In keeping with retry_connection_id being the SCID that is used on Retry packets, I have a preference for initial_connection_id, since it is the SCID used on Initial packets. 

In the same vein, and this is a weaker preference, I would propose changing original_connection_id to original_initial_connection_id. I think this makes it clear that this might have been initial_connection_id had it not been for a Retry.

File: draft-ietf-quic-transport.md [hunk 12]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
 
+## Authenticating Connection IDs {#cid-auth}
+
+The choice each endpoint makes about connection IDs during the handshake is
+authenticated by including all values in transport parameters; see
+{{transport-parameters}}. This ensures that all connection IDs are authenticated
+by the cryptographic handshake.
+
+Each endpoint includes the value of the Source Connection ID field from Initial
+packets it sends in the handshake_connection_id transport parameter; see
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:1486]:
```suggestion
Each endpoint includes the value of the Source Connection ID field from the 
most recent Initial packet it sent in the handshake_connection_id transport
parameter; see
```

**Comment by martinthomson** [comment]
I changed to initial_connection_id.  I don't necessarily like that we are using "original" and "initial", but I think that's marginally better than creating the obviously redundant "original_initial_" variant of this.

File: draft-ietf-quic-transport.md [hunk 13]
```diff
@@ -1544,7 +1588,6 @@ that are smaller than the remembered value of the parameters.
 * initial_max_stream_data_uni
 * initial_max_streams_bidi
 * initial_max_streams_uni
-* active_connection_id_limit
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md:1547]: Why was this removed in this PR?

Comment (martinthomson) [reply draft-ietf-quic-transport.md:1547]: A rebasing error.  I added something to this list, sorted it, removed it, now fixed it.

File: draft-ietf-quic-transport.md [hunk 14]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
+
+* absence of the retry_connection_id transport parameter from the server after
+  receiving a Retry packet,
+
+* presence of the retry_connection_id transport parameter when no Retry packet
+  was received, or
+
+* a mismatch between values received from a peer in these transport parameters
+  and the value sent in the corresponding Destination Connection ID fields of
+  Initial packets.
```
Comment (MikeBishop) [inline draft-ietf-quic-transport.md:1515]:
```suggestion
  Initial packets or the Source Connection ID field of a Retry packet.
```
...though perhaps this is fine, since the client will then use this value as its DCID for the next Initial packet.

Comment (martinthomson) [reply draft-ietf-quic-transport.md:1515]: > the client will then use this value as its DCID for the next Initial

Yeah, that was my thought.  The Retry.SCID thing is a choice people are free to use, but that doesn't need to be written down.  This way is simpler.

File: draft-ietf-quic-transport.md [hunk 15]
```diff
@@ -4609,10 +4662,9 @@ The following transport parameters are defined:
 original_connection_id (0x00):
 
 : The value of the Destination Connection ID field from the first Initial packet
-  sent by the client.  This transport parameter is only sent by a server.  This
-  is the same value sent in the "Original Destination Connection ID" field of a
-  Retry packet (see {{packet-retry}}).  A server MUST include the
-  original_connection_id transport parameter if it sent a Retry packet.
+  sent by the client; see {{cid-auth}}.  This transport parameter is only sent
+  by a server.  The value of this parameter depends on whether the server
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:4666]: I don't understand the sentence "The value of this parameter depends on whether the server sent a Retry packet". From reading the rest of this PR it sounds like this TP has the same value no matter what happened with RETRY. Am I missing something?

File: draft-ietf-quic-transport.md [hunk 16]
```diff
@@ -4786,17 +4838,28 @@ active_connection_id_limit (0x0e):
   When a zero-length connection ID is being used, the active_connection_id_limit
   parameter MUST NOT be sent.
 
+initial_connection_id (0x0f):
+
+: The value that the endpoint included in the Source Connection ID field of the
+  first Initial packet it sends during the handshake; see {{cid-auth}}.
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:4844]: nit: I would s/packet it sends during the handshake/packet it sent on this connection/

File: draft-ietf-quic-transport.md [hunk 17]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
 
 
+## Authenticating Connection IDs {#cid-auth}
+
+The choice each endpoint makes about connection IDs during the handshake is
+authenticated by including all values in transport parameters; see
+{{transport-parameters}}. This ensures that all connection IDs used for the
+handshake are also authenticated by the cryptographic handshake.
+
+Each endpoint includes the value of the Source Connection ID field from the most
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:1485]: Should this be the "first packet it sent" as opposed to "most recent" for consistency?

File: draft-ietf-quic-transport.md [hunk 18]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
+recent Initial packet it sent in the initial_connection_id transport
+parameter; see {{transport-parameter-definitions}}. A server includes the
+Destination Connection ID field it receives in original Initial packets from the
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:1488]: I would rephrase this to "A server includes the Destination Connection ID field from the first Initial packet it received from the client in the original_connection_id transport parameter; note that if the server sent a Retry packet this refers to the first Initial packet received before sending the Retry packet."

File: draft-ietf-quic-transport.md [hunk 19]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
+client - Initial packets received by the server prior to sending a Retry packet
+- in the original_connection_id transport parameter. After sending a Retry
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:1490]: I would s/After sending a/If it sent a/

File: draft-ietf-quic-transport.md [hunk 20]
```diff
@@ -1475,6 +1475,49 @@ lifetime of a connection, especially in response to connection migration
...
+packet, a server also includes the Source Connection ID field from the Retry
+packet in the retry_connection_id transport parameter.
+
+The values provided by a peer for these transport parameters MUST match the
+values that an endpoint used in the Destination Connection ID field of Initial
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:1495]: This sentence appears to apply to all 3 transport parameters, but it mentions Destination CID. I would s/Destination/Source and Destination/

File: draft-ietf-quic-transport.md [hunk 21]
```diff
@@ -4395,13 +4439,22 @@ A client MUST NOT reset the packet number for any packet number space after
...
-original_connection_id transport parameter (see
-{{transport-parameter-definitions}}).  If the server sends a Retry packet, it
-MUST include the Destination Connection ID field from the client's first
-Initial packet in the transport parameter.
+retry_connection_id transport parameter; see
+{{transport-parameter-definitions}}.  If it sends a Retry packet, the server
+also subsequently includes the value of the Source Connection ID field from
+the Retry packet in its retry_connection_id transport parameter.
+
+A server always includes an original_connection_id transport parameter. If it
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:4447]: I think we can remove this entire paragraph as it appears to be redundant with other parts of this PR, and it seems to have discrepancies.

File: draft-ietf-quic-transport.md (hunk 16 above)
Comment (huitema) [reply draft-ietf-quic-transport.md:4844]: Everywhere else, "initial_connection_id" is defined as _The value that the endpoint included in the  **Destination** Connection ID field of the first Initial packet it sends during the handshake._
I think that the simplest solution is to s/Source/Destination/

File: draft-ietf-quic-transport.md (hunk 18 above)
Comment (huitema) [reply draft-ietf-quic-transport.md:1488]: We already have a definition for an original_connection_id transport parameter (see section  17.2.5.3. of the transport spec, Continuing a Handshake After Retry). From the point of view of the client, we have the following:

1) Without retry:
* Initial DCID: DCID of the first Initial packet sent by the client.
* Server SCID: CID chosen by the server to identify the connection during the handshake. Also used by the client for further Initial or Handshake packets once the client receives a first handshake message from the server.

2) With Retry:
* Original DCID: the DCID in the first Initial packet sent by the client, when that packet triggered a Retry.
* Server suggested DCID: the SCID of the Retry packet sent by the server. 
* Initial DCID: the DCID in the first Initial packet sent by the client in response to the Retry. MUST be set to the Server suggested DCID.
* Server SCID: the SCID chosen by the server to identify the connection. May or may not be the same as Server suggested DCID.

Maybe we should tighten these definitions.

File: draft-ietf-quic-transport.md (hunk 16 above)
Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md:4844]: I don't think that's right @huitema. `initial_connection_id` refers to the Source Connection ID because that's the one that needs protection. The Destination Connection ID is covered by `original_connection_id `.

File: draft-ietf-quic-transport.md (hunk 18 above)
Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md:1488]: `original_connection_id` used to only apply when a Retry was involved. This PR changes that: servers now always send `original_connection_id` no matter what.

File: draft-ietf-quic-transport.md (hunk 16 above)
Comment (huitema) [reply draft-ietf-quic-transport.md:4844]: @DavidSchinazi If you want to refer to the source CID set by the client in its first Initial packet, you should really use a different name than `initial_connection_id`, because this is widely used to refer to the initial destination CID. Maybe you should call that `initial_source_connection_id`. But using the same name for 2 concepts seems bad.

Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md:4844]: I didn't pick the name, but I agree that `initial_source_connection_id` might be clearer than `initial_connection_id`.

File: draft-ietf-quic-transport.md [hunk 22]
```diff
@@ -1503,7 +1548,7 @@ An endpoint MUST NOT send a parameter more than once in a given transport
 parameters extension.  An endpoint SHOULD treat receipt of duplicate transport
 parameters as a connection error of type TRANSPORT_PARAMETER_ERROR.
 
-A server MUST include the original_connection_id transport parameter
+A server MUST include the original_destination_connection_id transport parameter
```
Comment (marten-seemann) [inline draft-ietf-quic-transport.md:1551]: This is inconsistent with the error condition above:

> * absence of the original_destination_connection_id transport parameter from the server"

**Comment by martinduke** [comment]
Is it not necessary to authenticate both client CIDs in a Retry scenario?

C->S: Initial SCID = C1 (attacker modifies to C2), DCID = S1
C<-S: Retry SCID = S2, DCID = C1
C->S: Initial SCID = C1, DCID = S2
C<-S: Initial SCID = S3, DCID = C1
....
C->S: Short DCID = S3
C<-S: Short DCID = C1

Admittedly, I don't really understand the threat model here, but this isn't authenticate unless the Initial SCID is in the Token. I don't think a TP is needed, as the server can reject the 2nd initial if there's a mismatch in SCID.

Relatedly, if it's written anywhere that the client MUST NOT change its CID during a Retry, it's not in a place that's obvious to me.

File: draft-ietf-quic-transport.md [hunk 23]
```diff
@@ -1475,6 +1475,51 @@ lifetime of a connection, especially in response to connection migration
...
+includes the Source Connection ID field from the Retry packet in the
+retry_source_connection_id transport parameter.
+
+The values provided by a peer for these transport parameters MUST match the
+values that an endpoint used in the Destination and Source Connection ID fields
+of Initial packets that it sent. Including connection ID values in transport
+parameters and verifying them ensures that that an attacker cannot influence
+the choice of connection ID for a successful connection by injecting packets
+carrying attacker-chosen connection IDs during the handshake. An endpoint MUST
+treat any of the following as a connection error of type PROTOCOL_VIOLATION:
```
[the remark this answers is not in the document]
Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md:1501]: I'm not sure I agree. TRANSPORT_PARAMETER_ERROR indicates a parse issue, whereas this validation is performed somewhere else in our implementation.

**Comment by martinthomson** [comment]
@martinduke, if the attacker modifies the SCID in an Initial that ultimately causes a Retry to be sent, that doesn't affect the final state in any way, so there is no harm done.

The reason we authenticate the DCID on that packet in the way we do is to verify that the the server is cooperating with the entity that generated the Retry token.  Also, the client-chosen DCID value might persist as the server could use that value to determine its choice of connection ID.  The client can't change its mind, so authenticating a single value is enough.

That said, I will take on advisement that we don't require that the client doesn't change its connection ID.  Issue inbound.

**Comment by martinthomson** [comment]
I just added a picture.

This new figure shows the complete set of messages that determine the
connection ID that each peer uses.  The explanation describes how the
connection IDs in those messages are included in transport parameters.

Hopefully this aids in understanding how this is put together.  It's not
complicated, but it's not necessarily clear from reading a set of rules
that this is the case.

File: draft-ietf-quic-transport.md (hunk 22 above)
Comment (martinthomson) [reply draft-ietf-quic-transport.md:1647]: Not following.  These are consistent by my reading:

> A server MUST include the original_destination_connection_id transport parameter

> An endpoint MUST
treat any of the following as a connection error of type PROTOCOL_VIOLATION:
> * absence of the original_destination_connection_id transport parameter from the server

Comment (DavidSchinazi) [reply draft-ietf-quic-transport.md:1647]: @martinthomson you truncated the sentence in your quote. The full sentence from your PR is:

> A server MUST include the original_destination_connection_id transport parameter
> ({{transport-parameter-definitions}}) if it sent a Retry packet to enable
> validation of the Retry, as described in {{packet-retry}}.

How about we keep the first line, but replace the rest:

> A server MUST include the original_destination_connection_id transport parameter
> to mitigate an attacker's ability to tamper with connection IDs during the handshake.

Comment (martinthomson) [reply draft-ietf-quic-transport.md:1647]: Ahh, that's lazy of me.  Yes, that is good.  Thanks.

Comment (martinthomson) [reply draft-ietf-quic-transport.md:1647]: Oh, that's a throwback paragraph.  I'm going to replace that with a more generic reference.

File: draft-ietf-quic-transport.md [hunk 24]
```diff
@@ -1542,6 +1542,80 @@ lifetime of a connection, especially in response to connection migration
...
+                                  <- Initial: DCID=C1, SCID=S3
+                             ...
+1-RTT: DCID=S3 ->
+                                             <- 1-RTT: DCID=C1
+~~~
+{: #fig-auth-cid title="Use of Connection IDs in a Handshake"}
+
+For the handshake in {{fig-auth-cid}} the client sets the value of the
+initial_source_connection_id transport parameter to `C1`. If the server sends a
+Retry packet (that is, the packets marked with a '*' are sent), it sets
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:1609]: I really like the diagram, but I would prefer to split it in two. There are two possible handshakes described here: without and with retry involved. I think that the `the packets marked with a '*' are sent` caveat is confusing, because when those packets aren't sent, the `1-RTT: DCID=S3` needs to be replaced by `1-RTT: DCID=S2`. It might be simpler to have two diagrams and two examples.

File: draft-ietf-quic-transport.md [hunk 25]
```diff
@@ -4422,15 +4498,16 @@ A client MUST NOT reset the packet number for any packet number space after
...
+the Retry packet in its retry_source_connection_id transport parameter.
 
 If the client received and processed a Retry packet, it MUST validate that the
-original_connection_id transport parameter is present and correct; otherwise, it
-MUST validate that the transport parameter is absent.  A client MUST treat a
-failed validation as a connection error of type TRANSPORT_PARAMETER_ERROR.
+retry_source_connection_id transport parameter is present and correct;
+otherwise, it MUST validate that the transport parameter is absent. A client
+MUST treat a failed validation as a connection error of type
+TRANSPORT_PARAMETER_ERROR.
```
Comment (DavidSchinazi) [inline draft-ietf-quic-transport.md:4510]: This contradicts line 1568
> An endpoint MUST treat any of the following as a connection error of type PROTOCOL_VIOLATION

I have a slight preference for PROTOCOL_VIOLATION over TRANSPORT_PARAMETER_ERROR, though I don't much care as long as this section matches the other one.

File: draft-ietf-quic-transport.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (marten-seemann) [reply draft-ietf-quic-transport.md:1568]: Our rule for FRAME_ENCODING_ERROR is that we use that error code when the error is detectable while parsing the frame, without accessing connection state. I'd argue that we should apply a similar logic here. The absence of `initial_source_connection_id` and `original_destination_connection_id` definitely falls in that category, and I'd put the absence / presence of `retry_source_connection_id` in the same category, although I could see why someone could make an argument against this.

The mismatch seems to be a clear case for PROTOCOL_VIOLATION.

[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-transport.md:1568]: I like that rationale.  I could also see my way to choose TRANSPORT_PARAMETER_ERROR.  We use that for transport parameters that are present when they are disallowed (like preferred_address from a client) in addition to obvious encoding problems, which suggests that you might implement this as `if transport_parameters.bad() then TRANSPORT_PARAMETER_ERROR`, but this validation does involve accessing external state, as you say, so a different code is fully justified.
